File: src/day7/day7.py
class TreeNode():
    def __init__(self):
        self.title = ""
        self.size = 0
        self.children = []
        self.parent = None
        self.type = ""

def dfs_print(node, offset = ""):
    if not node:
        return
    
    print(offset + node.title + " (" + str(node.size) + ")")

    for child in node.children:
        dfs_print(child, offset+" ")

def calc_dir_sizes(node):
    if not node:
        return 0
    
    if node.type == "dir":
        s = 0
        for c in node.children:
            s += calc_dir_sizes(c)
        node.size = s
        return s
    else:
        return node.size

def dfs_sum(node):
    if not node:
        return 0
    if node.type != "dir":
        return 0

    s = node.size if node.size <= 100000 else 0
    for c in node.children:
        s += dfs_sum(c)
    return s

def dfs_smallest(node, used):
    if not node:
        return  float("inf")
    if node.type != "dir":
        return  float("inf")

    free = 70000000 - used
    need = 30000000 - free

    m = node.size if node.size > need else float("inf")

    for c in node.children:
        m = min(m, dfs_smallest(c, used))

    return m

def main():

    cd = TreeNode()
    cd.type = "dir"

    root = TreeNode()
    root.title = "/"
    root.parent = cd
    root.type = "dir"

    head = root

    cd.children.append(root)
    
    with open("src/day7/input.txt", "r") as f:        
        for line in f:

            split = line.split(" ")
            split = [x.strip() for x in split]

            if split[0] == "$":
                
                if split[1] == "cd":

                    if split[2] == "..":
                        cd = cd.parent

                    else:

                        for child in cd.children:

                            if child.title == split[2]:
                                cd = child

            elif split[0] == "dir":
                node = TreeNode()
                node.title = split[1]
                node.parent = cd
                node.type = "dir"
                cd.children.append(node)

            else:
                node = TreeNode()
                node.title = split[1]
                node.parent = cd
                node.size = int(split[0])
                node.type = "file"
                cd.children.append(node)

    calc_dir_sizes(head)

    print("Day7 Part1: " + str(part1(head)))
    print("Day7 Part2: " + str(part2(head)))


def part1(head):
    return dfs_sum(head)

def part2(head):
    return dfs_smallest(head, head.size)

if __name__ == "__main__":
    main()

File: src/day7/day7.rs
// use std::{fs, fmt};
// use std::cell::RefCell;
// use std::rc::Rc;

// #[derive(PartialEq)]
// struct TreeNode {
//   pub value: Option<File>,
//   pub children: Vec<Rc<RefCell<TreeNode>>>,
//   pub parent: Option<Rc<RefCell<TreeNode>>>,
// }

// impl TreeNode {
//   pub fn new() -> TreeNode {
//     return TreeNode {
//       value: None,
//       children: vec![],
//       parent: None,
//     };
//   }

//   pub fn add_child(&mut self, new_node: Rc<RefCell<TreeNode>>) {
//     self.children.push(new_node);
//   }

//   pub fn print(&self) -> String {
//     if let Some(value) = self.value {
//       return value.to_string();
//     } else {
//       return String::from("[")
//         + &self
//           .children
//           .iter()
//           .map(|tn| tn.borrow().print())
//           .collect::<Vec<String>>()
//           .join(",")
//         + "]";
//     }
//   }
// }

// #[derive(PartialEq)]
// enum File {

//     File((String, u32)),
//     Dir(String),

// }

// impl fmt::Display for File {

//     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
//         match *self {
//             File::File((title, size)) => write!(f, "{title}, {size}"),
//             File::Dir(title) => write!(f, "{title}"),
//         }
//      }

// }

// pub fn run() -> (u32, u32) {

//     let input = fs::read_to_string("src/day7/test.txt")
//     .expect("Should have been able to read the file");

//     let p1 = part1(&input);
//     let p2 = part2(&input);

//     return (p1, p2);

// }

// fn part1(input: &str) -> u32 {

//     let mut current_dir = Rc::new(RefCell::new(TreeNode::new()));

//     let root = Rc::new(RefCell::new(TreeNode::new()));
//     root.borrow_mut().value = Some(File::Dir("/".to_string()));


//     for line in input.lines() {

//         let command: Vec<&str> = line.split(" ").collect();

//         let first = *command.iter().next().unwrap();
//         let second = *command.iter().nth(1).unwrap();

//         match first {

//             "$" => {
                
//                 match second {

//                     "cd" => {
                        
//                         for node in &current_dir.borrow().children {

//                             match &node.borrow().value {

//                                 Some(File::File((_, _))) => {}
//                                 Some(File::Dir(title)) => {

//                                     current_dir = *node;

//                                 }
//                                 None => {}

//                             }

//                         } 

//                     },
//                     "ls" => {}
//                     _ => {}

//                 }

//             },
//             "dir" => {println!("Dir")}
//             _ => {println!("File")}

//         };

//     }
    
//     0

// }

// fn part2(input: &str) -> u32 {

//     0      

// }